---
title: "Figure 1 PCA Plots"
author: "Sara Gosline"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(synapser)
synLogin()
require(tidyverse)


```

## Table 1
To generate table 1 we need to collect all mRNA-seq and genomic sequencing data we have to summarize.

```{r rna-Seq, warning=FALSE}
tabids<-synTableQuery('select distinct tableId from syn21221980')$asDataFrame()

vars="specimenID,individualID,Symbol,totalCounts,zScore,tumorType,nf1Genotype,sex,isCellLine,transplantationType,study"
full.tab<-do.call(rbind,lapply(tabids$tableId,function(x) synTableQuery(paste('select',vars,'from',x))$asDataFrame()))

#full.tab<-full.tab
#%>%
red.tab<-subset(full.tab,tumorType%in%c('Malignant peripheral nerve sheath tumor','Neurofibroma','Malignant Peripheral Nerve Sheath Tumor','Plexiform Neurofibroma','Cutaneous Neurofibroma'))%>%
    subset(isCellLine%in%c(NA,"FALSE"))%>%
  subset(is.na(transplantationType))

red.tab$tumorType<-gsub("Malignant peripheral nerve sheath tumor",'Malignant Peripheral Nerve Sheath Tumor',red.tab$tumorType)


```

Now we have to collect the exome seq and WGS data. We have two tables containing gene variant data. All located at a table in `syn21266269`.

```{r variant data,warning=FALSE}
tabids<-synTableQuery('select distinct tableId from syn21266269')$asDataFrame()

vars="Hugo_Symbol,Protein_position,specimenID,IMPACT,FILTER,ExAC_AF,gnomAD_AF"
exome.tab<-do.call(rbind,lapply(tabids$tableId,function(x) synTableQuery(paste('select',vars,'from',x))$asDataFrame()))%>%
   subset(gnomAD_AF<0.01)

```

Now we want to count how many samples of each tumor type and whether or not they have sequencing or transcript levels

```{r table 1}
samp.data<-red.tab%>%
  select(individualID,specimenID,tumorType,sex,study)%>%
  distinct()%>%
  mutate(hasGenomicData=specimenID%in%exome.tab$specimenID)

samp.data$sex<-gsub('female','Female',gsub('^male','Male',samp.data$sex))
samp.data$study[is.na(samp.data$study)]<-'CBTTC'

#set this up to be the annotation for the heatmaps
rownames(samp.data)<-samp.data$specimenID

samps<-samp.data%>%
  group_by(tumorType)%>%
  summarize(individuals=n_distinct(individualID),samples=n_distinct(specimenID),numGenomic=length(which(hasGenomicData)))

samp.data<-apply(samp.data,2,as.factor)%>%
  as.data.frame()%>%
  select(sex,study,hasGenomicData,tumorType)

##save mapping for later
tumType<-samp.data%>%
  rownames_to_column('sample')%>%
  select(sample,tumorType,study)%>%
  distinct()

DT::datatable(samps)

```

## Figure 1 Panel A

We can create PCA plots of the samples. 

```{r transcript heatmap, echo=FALSE}
library(ggfortify)


as.mat<-red.tab%>%
  reshape2::acast(Symbol~specimenID,value.var='zScore',fun.aggregate = function(x) mean(x,na.rm=T))%>%
  as.matrix()
as.mat[which(!is.finite(as.mat),arr.ind = T)]<-0.0

comps<-prcomp(t(as.mat))

dist.vals<-as.dist(comps$x)%>%
  as.matrix()%>%abs()%>%
  as.data.frame()%>%
  rownames_to_column('from')%>%
  pivot_longer(cols=(-from),names_to='to')

gene.tt.df<-dist.vals%>%
  left_join(rename(tumType,from=sample,fromType=tumorType),by='from')%>%
  left_join(rename(tumType,to=sample,toType=tumorType),by='to')%>%
  filter(fromType==toType)%>%
  rowwise()%>%
  mutate(fromto=paste(sort(c(from,to)),collapse=','))%>%
  select(fromto,value,fromType)%>%distinct()
#  group_by(fromType)%>%summarize(meanDist=mean(value))

#DT::datatable(lv.tt.df)

gene.study.df<-dist.vals%>%
  left_join(rename(tumType,from=sample,fromStudy=study),by='from')%>%
  left_join(rename(tumType,to=sample,toStudy=study),by='to')%>%
  filter(fromStudy==toStudy)%>%
    rowwise()%>%
  mutate(fromto=paste(sort(c(from,to)),collapse=','))%>%
  select(fromto,value,fromStudy)%>%distinct()



autoplot(comps,data=samp.data,colour="tumorType",shape='study')+ggtitle('RNAseq-derived clustering')+theme_bw()+scale_color_manual(values = c("Cutaneous Neurofibroma"="#ca054d", 
                               "Plexiform Neurofibroma" = "#3b1c32",
                               "Neurofibroma" = "#a4d4b4",
                               "Malignant Peripheral Nerve Sheath Tumor" = "#ffcf9c"))


ggsave('fig1panelA.pdf',useDingbats=FALSE)
#dev.off()

```


## Figure 1 Panel C

Now we can do the same for latent variables to see if the clustering is any more coherent.

```{r LV plots,echo=FALSE}

mvscores=synTableQuery("select * From syn21046991 where isCellLine <> TRUE")$asDataFrame()
with.var<-mvscores%>%
    group_by(latent_var)%>%
    mutate(lvVar=var(value))%>%
    arrange(desc(lvVar))%>%
    dplyr::select(latent_var,lvVar)%>%distinct()

as.mat<-mvscores%>%subset(specimenID%in%rownames(samp.data))%>%
  reshape2::acast(latent_var~specimenID,value.var='value',fun.aggregate = mean)%>%
  as.matrix()

comps<-prcomp(t(as.mat))

dist.vals<-as.dist(comps$x)%>%
  as.matrix()%>%abs()%>%
  as.data.frame()%>%
  rownames_to_column('from')%>%
  pivot_longer(cols=(-from),names_to='to')

lv.tt.df<-dist.vals%>%
  left_join(rename(tumType,from=sample,fromType=tumorType),by='from')%>%
  left_join(rename(tumType,to=sample,toType=tumorType),by='to')%>%
  filter(fromType==toType)%>%
  rowwise()%>%
  mutate(fromto=paste(sort(c(from,to)),collapse=','))%>%
  select(fromto,value,fromType)%>%distinct()
#  group_by(fromType)%>%summarize(meanDist=mean(value))

#DT::datatable(lv.tt.df)

lv.study.df<-dist.vals%>%
  left_join(rename(tumType,from=sample,fromStudy=study),by='from')%>%
  left_join(rename(tumType,to=sample,toStudy=study),by='to')%>%
  filter(fromStudy==toStudy)%>%
    rowwise()%>%
  mutate(fromto=paste(sort(c(from,to)),collapse=','))%>%
  select(fromto,value,fromStudy)%>%distinct()
#  group_by(fromStudy)%>%summarize(meanDist=mean(value))

#DT::datatable(lv.study.df)


autoplot(comps,data=samp.data,colour="tumorType",shape='study')+ggtitle('LV-derived clustering')+theme_bw()+scale_color_manual(values = c("Cutaneous Neurofibroma"="#ca054d", 
                               "Plexiform Neurofibroma" = "#3b1c32",
                               "Neurofibroma" = "#a4d4b4",
                               "Malignant Peripheral Nerve Sheath Tumor" = "#ffcf9c"))


ggsave('fig1panelB.pdf',useDingbats=FALSE)
#dev.off()


```

## Summary stats

We can summarize the distances between points.

```{r summary stats}

tt.tab<-gene.tt.df%>%
  select(fromto,value)%>%
    rename(genes=value)%>%
    left_join(lv.tt.df,by='fromto')%>%
    rename(LVs=value,tumorType=fromType)%>%
  pivot_longer(cols=c(genes,LVs),names_to='Values Clustered')

study.tab<-gene.study.df%>%
  select(fromto,value)%>%
    rename(genes=value)%>%
    left_join(lv.study.df,by='fromto')%>%
    rename(LVs=value,study=fromStudy)%>%
  pivot_longer(cols=c(genes,LVs),names_to='Values Clustered')

full.tab<-rbind(tt.tab%>%select(fromto,`Values Clustered`,`Pairwise Distance`=value,Group=tumorType)%>%mutate(`Distance Measured`='Within Tumor Type'),study.tab%>%select(fromto,`Values Clustered`,`Pairwise Distance`=value,Group=study)%>%mutate(`Distance Measured`='Batch'))

sigs<-full.tab%>%group_by(`Distance Measured`)%>%pivot_wider(names_from='Values Clustered',values_from='Pairwise Distance')%>%group_by(`Distance Measured`)%>%mutate(pval=t.test(genes,LVs)$p.value)%>%select(`Distance Measured`,pval)%>%distinct()

DT::datatable(sigs)

ggplot(subset(full.tab,`Distance Measured`=='Within Tumor Type'))+geom_boxplot(aes(y=`Pairwise Distance`,x=`Group`,fill=`Values Clustered`))+scale_y_log10()+theme_bw()+scale_fill_brewer(palette='Dark2')+ theme(axis.text.x=element_text(angle=45, hjust=1))
ggsave('fig_s2_withinTumorLVclustering.pdf')
```

## Clustering based on top 40 LVs

When we cluster based on only those LVs selected by the RF we do better.

```{r updated lvs}
mvscores=synTableQuery("select * From syn21046991 where isCellLine <> TRUE")$asDataFrame()
with.var<-mvscores%>%
    group_by(latent_var)%>%
    mutate(lvVar=var(value))%>%
    arrange(desc(lvVar))%>%
    dplyr::select(latent_var,lvVar)%>%distinct()

selectedlvs=synTableQuery("select latent_var from syn21449221")$asDataFrame()%>%
  select(latent_var)%>%
  distinct()

as.mat<-mvscores%>%subset(specimenID%in%rownames(samp.data))%>%
  subset(latent_var%in%selectedlvs$latent_var)%>%
  reshape2::acast(latent_var~specimenID,value.var='value',fun.aggregate = mean)%>%
  as.matrix()

comps<-prcomp(t(as.mat))

dist.vals<-as.dist(comps$x)%>%
  as.matrix()%>%abs()%>%
  as.data.frame()%>%
  rownames_to_column('from')%>%
  pivot_longer(cols=(-from),names_to='to')

joined.df<-dist.vals%>%
  left_join(rename(tumType,from=sample,fromType=tumorType),by='from')%>%
  left_join(rename(tumType,to=sample,toType=tumorType),by='to')%>%
  filter(fromType==toType)

top.lv.dist<-joined.df%>%group_by(fromType)%>%summarize(meanDist=mean(value))


DT::datatable(top.lv.dist)

autoplot(comps,data=samp.data,colour="tumorType",shape='study')+ggtitle('LV-derived clustering')+theme_bw()+scale_color_manual(values = c("Cutaneous Neurofibroma"="#ca054d", 
                               "Plexiform Neurofibroma" = "#3b1c32",
                               "Neurofibroma" = "#a4d4b4",
                               "Malignant Peripheral Nerve Sheath Tumor" = "#ffcf9c"))


ggsave('fig1panel_alt.pdf',useDingbats=FALSE)
```
