---
title: "Figure 3 ideas"
author: "Sara Gosline"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(synapser)
synLogin()
require(tidyverse)


```

## Grab data from table
The goal of this figure is to summarize the results in `syn21326918`.

```{r rna-Seq, warning=FALSE}

rf.results<-synTableQuery('select * from syn21326918')$asDataFrame()%>%
  select(-c(ROW_ID,ROW_VERSION))

DT::datatable(rf.results)

```

## Figure 3A

Now we can go ahead and plot a venn diagram of the RF results. The first thing we want to do is evaluate how many latent variables predict which tumor types. This Venn Diagram illustrates that there are numerous LVs that predict many different tumor types. 

```{r Venn,warning=FALSE}
library(ggvenn)


ggvenn(data.frame(rf.results$latent_var,apply(rf.results[,2:5],2,function(x) ifelse(is.na(x),F,T))),fill_color = c('#ca054d', '#ffcf9c', '#a4d4b4', '#3b1c32'))+theme_bw()



pdf('vennDiag.pdf')

ggvenn(data.frame(rf.results$latent_var,apply(rf.results[,2:5],2,function(x) ifelse(is.na(x),F,T))), 
  fill_color = c('#ca054d', '#ffcf9c', '#a4d4b4', '#3b1c32'))+theme_bw()
dev.off()
```

## Figure 3B
Now we want to count how many samples of each tumor type and whether or not they have sequencing or transcript levels. This plot is less successful and I imagine we'll drop it. The idea is to show which variables have 'correlates' - either immune cell types, metaviper proteins, or gene mutations.

```{r barpots}
library(ggplot2)

tdf<-rf.results%>%
    pivot_longer(c(cNF,MPNST,NF,pNF),names_to='predictiveOf',values_drop_na = TRUE)%>%
    pivot_longer(c("Proteins with correlated activity","Immune Signatures","Number Genes Mutated"),values_to='CorrelatedSignals',values_drop_na = FALSE)%>%
  rename(`Correlated Features`='name')

#ggplot(tdf)+geom_bar(aes(x=`Correlated Features`,fill=predictiveOf),position='dodge')+ggtitle('Number of variables with correlated features')

scale_fill_venn <- function(...){
    ggplot2:::manual_scale(
        'color', 
        values = setNames(c('#ca054d', '#ffcf9c', '#a4d4b4', '#3b1c32'), as.factor(c('cNF','MPNST','NF','pNF'))),
        ...
    )
}

##trying to re-create the venn diagram colors
ggplot(tdf,aes(y=latent_var,shape=`Correlated Features`,x=CorrelatedSignals,color=predictiveOf,alpha=0.5))+geom_point()+scale_x_log10()+scale_fill_venn()+theme_bw()

pdf('fig3dotplot.pdf')
ggplot(tdf,aes(y=latent_var,shape=`Correlated Features`,x=CorrelatedSignals,color=predictiveOf,alpha=0.5))+geom_point()+scale_x_log10()+scale_fill_venn()+theme_bw()
dev.off()

```

# Alternate Figure 3B

Perhaps a heatmap would be better. I tried to plot it, see attached. The idea is to show which latent variables predict tumor type (red), and how many of those correlates described above exist. Blue indicates 0/NA.

```{r fig3b again, warning=FALSE,echo=FALSE}
library(pheatmap)
urf<-rf.results%>%column_to_rownames('latent_var')%>%
    apply(2,function(x) as.numeric(ifelse(is.na(x),-1000,ifelse(x=='X',1000,x))))
rownames(urf)<-rf.results$latent_var

pheatmap(urf,cellwidth = 10,cellheight = 10,cluster_cols = F,clustering_method = 'ward', gaps_col=4, filename = 'fig3heatmap.pdf')

pheatmap(urf,cellwidth = 10,cellheight = 10,cluster_cols = F,gaps_col=4, clustering_method = 'ward')
```
