---
title: "Plotting LVs and Genetic Variants"
author: "Sara Gosline"
date: "`r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Get LVs selected by random forest and multiPLIER expression results, and gene variants and . 

```{r download data, echo=FALSE,warning=FALSE}
library(tidyverse)
library(synapser)

synLogin()


mp_res <- mp_res <- synTableQuery("select * from syn21046991")$filepath %>% read_csv() %>% filter(!is.na(tumorType))

##now get the genomic variant data
wgs.vars=synTableQuery("SELECT Hugo_Symbol,Protein_position,specimenID,IMPACT,FILTER,ExAC_AF,gnomAD_AF FROM syn20551862")$asDataFrame()
exome.vars=synTableQuery("SELECT Hugo_Symbol,Protein_position,specimenID,IMPACT,FILTER,ExAC_AF,gnomAD_AF FROM syn20554939")$asDataFrame()
all.vars<-rbind(select(wgs.vars,'Hugo_Symbol','Protein_position','specimenID','IMPACT','gnomAD_AF'),
    select(exome.vars,'Hugo_Symbol','Protein_position','specimenID','IMPACT','gnomAD_AF'))%>%
  subset(gnomAD_AF<0.01)

##EXPRESSION DATA
tabids<-synTableQuery('select distinct tableId from syn21221980')$asDataFrame()
vars="specimenID,individualID,Symbol,totalCounts,zScore,tumorType,nf1Genotype,sex"
full.tab<-do.call(rbind,lapply(tabids$tableId,function(x) synTableQuery(paste('select',vars,'from',x))$asDataFrame()))

#lets only get those genes that are expressed in all samples
expr.genes<-full.tab%>%group_by(Symbol)%>%
  summarize(minExpr=min(totalCounts))%>%
  subset(minExpr>0)%>%ungroup()%>%select(Symbol)%>%
  distinct()

top.lvs<-synTableQuery("SELECT * from syn21318452")$asDataFrame()


expr.vars<-subset(all.vars,Hugo_Symbol%in%expr.genes$Symbol)

samps<-intersect(mp_res$specimenID,expr.vars$specimenID)

mp_res<-mp_res%>%
  subset(specimenID%in%samps)

data.with.var<-mp_res%>%
  left_join(expr.vars,by='specimenID')

tums <- c("MPNST","NF", "pNF","cNF")
```

Now we filter for those genes with mutational status that significantly differentiates between LV values. 

```{r calculate variants,warning=FALSE,echo=FALSE}

top.genes=data.with.var%>%#group_by(tumorType)%>%
  mutate(numSamps=n_distinct(specimenID))%>%
      group_by(Hugo_Symbol)%>%
    mutate(numMutated=n_distinct(specimenID))%>%
    ungroup()%>%
  subset(numMutated>1)%>%
      subset(numMutated<(numSamps-1))%>%
  select(tumorType,Hugo_Symbol,numSamps,numMutated)%>%distinct()

vals<-data.with.var%>%subset(Hugo_Symbol%in%top.genes$Hugo_Symbol)%>%
    mutate(mutated=ifelse(is.na(IMPACT),'WT','Mutated'))%>%
  select(latent_var,tumorType,value,Hugo_Symbol,specimenID,mutated)%>%
  distinct()%>%
  spread(key=Hugo_Symbol,value='mutated',fill='WT')

##double check to make sure there are both mutated and unmutated values
counts<-vals%>%
  gather(key=gene,value=status,-c(latent_var,tumorType,value,specimenID))%>% 
    select(latent_var,tumorType,value,gene,specimenID,status)#%>%
  #  group_by(latent_var,gene)%>%
  #  mutate(numVals=n_distinct(status))%>%
#    mutate(numSamps=n_distinct(specimenID))%>%
  # subset(numVals==2)%>%ungroup()

#stored this to avoid recomputing
with.sig<-synTableQuery('select * from syn21407433')$asDataFrame()%>%
  subset(corP<0.01)

sig.vals<-with.sig%>%group_by(latent_var)%>%
    summarize(numGenes=n_distinct(gene))#%>%
  #subset(numGenes>1)

DT::datatable(sig.vals)

```

## LV Visualization

We have ~22 latent variables to showcase, and <10 genes that are associate with each. Most of these are actually the _same_ genes. 

First we just plot the number of mutations for each LV, colored by tumor Type.
```{r summary plot, warning=FALSE, echo=F}
mut.vals<-counts%>%
  subset(latent_var%in%with.sig$latent_var)%>%
  subset(status=='Mutated')%>%
  subset(gene%in%with.sig$gene)

require(ggplot2)
mut.counts=mut.vals%>%group_by(latent_var,gene,tumorType)%>%summarize(samplesMutated=n_distinct(specimenID))


library(viridis)
val<-viridis_pal()(length(unique(mut.vals$gene)))
names(val)<-unique(mut.vals$gene)

 ggplot(mut.vals,aes(x=latent_var,fill=gene))+geom_bar(position="dodge")+facet_grid(~tumorType)+theme_bw()+coord_flip()+scale_fill_manual(values=val)
 
  ggplot(mut.vals,aes(x=latent_var,fill=gene))+geom_bar(position="dodge")+facet_grid(~tumorType)+theme_bw()+coord_flip()+scale_fill_manual(values=val)
  
  ggsave('fig5_gv_lv_summary.pdf',useDingbats=FALSE)
```


## LV 851

This seems  like a good example of an 'almost' good LV
```{r plot LV, warning=FALSE, echo=FALSE}


library(viridis)
val<-viridis_pal()(2)


tplot<-subset(with.sig,latent_var=='LV 851')%>%
  left_join(subset(counts,latent_var=='LV 851'),by=c('gene','latent_var'))%>%
  select(gene,value,status,tumorType,specimenID)%>%
  distinct()%>%rename(`LV 851 score`=value)


p<-ggplot(tplot,aes(x=gene,y=`LV 851 score`,col=status))+
    geom_boxplot(outlier.shape=NA)+
    geom_point(position=position_jitterdodge(),aes(shape=tumorType,col=status,group=status))+
    theme_bw()+
    ggtitle(paste('LV 851 scores'))+
          scale_color_manual(values=val)
    print(p)
    
p
ggsave('fig5_lv851_example.pdf',useDingbats=F)
```

## LV 851 loadings

Lastly we want to get the loadings of LV 851 to compare. 

```{r plot other LV, warning=FALSE, echo=FALSE}
    

plier_model <- readr::read_rds(synGet("syn18689545")$path)
plier_loadings_df <- plier_model$Z %>% as.data.frame() %>% purrr::set_names(rownames(plier_model$B))

mp_res_subset <- mp_res %>% filter(latent_var=='LV 851')

i<-'LV 851'
  
  p1 <- ggplot(data = mp_res %>% filter(latent_var == i)) +
  ggbeeswarm::geom_quasirandom(aes(x=stringr::str_trunc(latent_var,30), y = value , color = tumorType, group = tumorType), dodge.width = 0.75) +
  theme_bw() +
  theme(axis.text.x = element_text(size = 10, angle = 0)) +
    labs(x = "", y = "multiPLIER value") +
    scale_color_manual(values = c("Cutaneous Neurofibroma"="#ca054d", 
                               "Plexiform Neurofibroma" = "#3b1c32",
                               "Neurofibroma" = "#a4d4b4",
                               "Malignant Peripheral Nerve Sheath Tumor" = "#ffcf9c"))
  
  tidy <- plier_loadings_df %>%
    dplyr::select(i) %>% 
    tibble::rownames_to_column('lv')
  p2 <- ggplot(tidy %>% top_n(30, get(i))) +
    geom_bar(aes(x=reorder(lv, -get(i)), y=get(i)), stat = "identity") +
    theme_bw() +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.5, size =6 )) +
      labs(x = "Gene", y = "LV Loading")
  
  print(p1)
  ggsave('fig5_lv851_expression.pdf',useDingbats=FALSE)
  print(p2)
    ggsave('fig5_lv851_loadings.pdf',useDingbats=FALSE)
  


```


```{r}
sessionInfo()
```
