---
title: "23 - RF-selected LV Genomic Variant Test"
author: "Sara Gosline"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(synapser)
synLogin()
library(tidyverse)
```

# Get Data

This is nearly entirely based on the code in notebook 09 and that in 11. 

We have latent variable expression analysis data
- [Latent Variable Table](https://www.synapse.org/#!Synapse:syn21046991/tables/)
- [Latent Variables selected by Random Forest](https://www.synapse.org/#!Synapse:syn21318452/tables/)

For this data we are also using any data for which there are gene variants (cNFs, pNFs, MPNSTs):
- [Exome-Seq variants](https://www.synapse.org/#!Synapse:syn20554939/tables/)
- [WGS Variants](https://www.synapse.org/#!Synapse:syn20551862/tables/) 
 
Lastly we need to filter by genes that are expressed to avoid getting too many un-qualifying variants.
- [RNA-Seq Data](https://www.synapse.org/#!Synapse:syn21221980/tables/)
 
Let's see if there are any LVs that split based on gene variant. Because we're having trouble scaling with the number of latent variables, I only look at variants that occur in less than 5% of the population. notice this is a difference from notebook #11.

```{r,warning=FALSE}
wgs.vars=synTableQuery("SELECT Hugo_Symbol,Protein_position,specimenID,IMPACT,FILTER,ExAC_AF,gnomAD_AF FROM syn20551862")$asDataFrame()
exome.vars=synTableQuery("SELECT Hugo_Symbol,Protein_position,specimenID,IMPACT,FILTER,ExAC_AF,gnomAD_AF FROM syn20554939")$asDataFrame()

all.vars<-rbind(select(wgs.vars,'Hugo_Symbol','Protein_position','specimenID','IMPACT','gnomAD_AF'),
    select(exome.vars,'Hugo_Symbol','Protein_position','specimenID','IMPACT','gnomAD_AF'))%>%
  subset(gnomAD_AF<0.01)

tabids<-synTableQuery('select distinct tableId from syn21221980')$asDataFrame()

vars="specimenID,individualID,Symbol,totalCounts,zScore,tumorType,nf1Genotype,sex"

full.tab<-do.call(rbind,lapply(tabids$tableId,function(x) synTableQuery(paste('select',vars,'from',x))$asDataFrame()))

#lets only get those genes that are expressed in all samples
expr.genes<-full.tab%>%group_by(Symbol)%>%
  summarize(minExpr=min(totalCounts))%>%
  subset(minExpr>0)%>%ungroup()%>%select(Symbol)%>%
  distinct()

top.lvs<-synTableQuery("SELECT * from syn21318452")$asDataFrame()

mp_res<-synTableQuery("SELECT * FROM syn21046991")$asDataFrame()%>%
  filter(isCellLine != "TRUE")%>%
  subset(latent_var%in%top.lvs$LatentVar)%>%
  select(latent_var,id,value,specimenID,tumorType,modelOf,diagnosis)

```

# Merge data together
For the purposes of this analysis we want to have only those samples wtih genomic data and only those latent variables that are selected by the Random Forest as predictive, and also those variants that are expressed.

```{r, warning=FALSE}

expr.vars<-subset(all.vars,Hugo_Symbol%in%expr.genes$Symbol)

samps<-intersect(mp_res$specimenID,expr.vars$specimenID)

mp_res<-mp_res%>%
  subset(specimenID%in%samps)#%>%
#  group_by(latent_var) %>%
#  mutate(sd_value = sd(value)) %>%
#  filter(sd_value > 0.025) %>%
#  ungroup()

```

# Retrieve Variant Data

Let's retrieve the LV data and summarize how many genes have mutations across samples.

```{r,warning=FALSE}
data.with.var<-mp_res%>%
  left_join(expr.vars,by='specimenID')

tab<-data.with.var

top.genes=tab%>%#group_by(tumorType)%>%
  mutate(numSamps=n_distinct(specimenID))%>%
      group_by(Hugo_Symbol)%>%
    mutate(numMutated=n_distinct(specimenID))%>%
    ungroup()%>%
  subset(numMutated>1)%>%
      subset(numMutated<(numSamps-1))%>%
  select(tumorType,Hugo_Symbol,numSamps,numMutated)%>%distinct()

gene.count=top.genes%>%group_by(tumorType)%>%
  mutate(numGenes=n_distinct(Hugo_Symbol))%>%
  mutate(minMutated=min(numMutated))%>%
  mutate(maxMutated=max(numMutated))%>%
  select(tumorType,numGenes,minMutated,maxMutated)%>%distinct()

DT::datatable(gene.count)

```
## Test significance of each gene/immune population

Now we can loop through every tumor type and gene with a Wilcoxon Rank Sum Test and correct for multiple testing for each LV.

```{r, warning=FALSE}
red.genes<-c("NF1","SUZ12","CDKN2A","EED")##for testing

##first spread the WT/Mutated values
vals<-tab%>%subset(Hugo_Symbol%in%top.genes$Hugo_Symbol)%>%
    mutate(mutated=ifelse(is.na(IMPACT),'WT','Mutated'))%>%
  select(latent_var,tumorType,value,Hugo_Symbol,specimenID,mutated)%>%
  distinct()%>%
  spread(key=Hugo_Symbol,value='mutated',fill='WT')

##double check to make sure there are both mutated and unmutated values
counts<-vals%>%
  gather(key=gene,value=status,-c(latent_var,tumorType,value,specimenID))%>% 
    select(latent_var,tumorType,value,gene,specimenID,status)%>%
    group_by(latent_var,gene)%>%
    mutate(numVals=n_distinct(status))%>%
    mutate(numSamps=n_distinct(specimenID))%>%
    subset(numVals==2)%>%ungroup()

#so now we have only 
with.sig<-counts%>%ungroup()%>%#subset(gene%in%top.genes$Hugo_Symbol)%>%
    group_by(latent_var,gene)%>%
  mutate(pval=wilcox.test(value~status)$p.value)%>%ungroup()%>%
  group_by(latent_var)%>%
  mutate(corP=p.adjust(pval))%>%ungroup()%>%
  select(latent_var,gene,pval,corP)%>%distinct()

sig.vals<-subset(with.sig,corP<0.01)

DT::datatable(sig.vals%>%group_by(latent_var)%>%summarize(numGenes=n_distinct(gene)))

```

Interesting! Some genes actually pass p-value correction. What do they look like? Here let's write the messiest possible code to print.

```{r,warning=FALSE}

for(ct in unique(sig.vals$latent_var)){
    tplot<-sig.vals[which(sig.vals$latent_var==ct),]
    if(nrow(tplot)==0)
      next
    
    print(ct)
    sigs=tplot%>%rowwise()%>%mutate(vals=paste(gene,format(corP,digits=3),sep=':'))%>%select(vals)%>%unlist()%>%paste(collapse=',')
    print(sigs)
        p<-counts%>%
    subset(latent_var==ct)%>%
    subset(gene%in%tplot$gene)%>%
    ggplot(aes(x=gene,y=value,col=status))+
    geom_boxplot(outlier.shape=NA)+
    geom_point(position=position_jitterdodge(),aes(shape=tumorType,group=status))+
    theme(axis.text.x = element_text(angle = 90, hjust = 1))+
    ggtitle(paste(ct,'scores\n',sigs))
#    if(method=='cibersort')
#      p<-p+scale_y_log10()
    print(p)
  }
#}


```


I'm not sure how to interpret this - it seems like most LVs have the same sets of genes that are mutated. Not sure why this is.


```{r, warning=FALSE}

#this is a failed attempt to group by tumor type
#with.sig<-counts%>%ungroup()%>%subset(gene%in%top.genes$Hugo_Symbol)%>%
#    group_by(latent_var,tumorType,gene)%>%
#  mutate(pval=t.test(value~status)$p.value)%>%
#  ungroup()%>%
#  group_by(latent_var)%>%
#  mutate(corP=p.adjust(pval))%>%ungroup()%>%
#  select(latent_var,tumorType,gene,pval,corP)%>%distinct()

#sig.vals<-subset(with.sig,corP<0.05)

#DT::datatable(sig.vals)
```
