---
title: "24 - RF-selected LV Table Statst"
author: "Sara Gosline"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(synapser)
synLogin()
library(tidyverse)
```

# Get Data

This is nearly entirely based on the code in notebook 09 and that in 11.

We have latent variable expression analysis data
- [Latent Variable Table](https://www.synapse.org/#!Synapse:syn21046991/tables/)
- [Latent Variables selected by Random Forest](https://www.synapse.org/#!Synapse:syn21318452/tables/)

For this data we are also using any data for which there are gene variants (cNFs, pNFs, MPNSTs):
- [Exome-Seq variants](https://www.synapse.org/#!Synapse:syn20554939/tables/)
- [WGS Variants](https://www.synapse.org/#!Synapse:syn20551862/tables/)

We need to filter by genes that are expressed to avoid getting too many un-qualifying variants.
- [RNA-Seq Data](https://www.synapse.org/#!Synapse:syn21221980/tables/)

Let's see if there are any LVs that split based on gene variant. Because we're having trouble scaling with the number of latent variables, I only look at variants that occur in less than 5% of the population. notice this is a difference from notebook #11.

## Get genomic data, calculate significant splits
```{r,warning=FALSE,echo=FALSE}

##variant data
wgs.vars=synTableQuery("SELECT Hugo_Symbol,Protein_position,specimenID,IMPACT,FILTER,ExAC_AF,gnomAD_AF FROM syn20551862")$asDataFrame()
exome.vars=synTableQuery("SELECT Hugo_Symbol,Protein_position,specimenID,IMPACT,FILTER,ExAC_AF,gnomAD_AF FROM syn20554939")$asDataFrame()

all.vars<-rbind(select(wgs.vars,'Hugo_Symbol','Protein_position','specimenID','IMPACT','gnomAD_AF'),
    select(exome.vars,'Hugo_Symbol','Protein_position','specimenID','IMPACT','gnomAD_AF'))%>%
  subset(gnomAD_AF<0.01)

#tables for expressed genes
tabids<-synTableQuery('select distinct tableId from syn21221980')$asDataFrame()
vars="specimenID,individualID,Symbol,totalCounts,zScore,tumorType,nf1Genotype,sex"

full.tab<-do.call(rbind,lapply(tabids$tableId,function(x) synTableQuery(paste('select',vars,'from',x))$asDataFrame()))

#lets only get those genes that are expressed in all samples
expr.genes<-full.tab%>%group_by(Symbol)%>%
  summarize(minExpr=min(totalCounts))%>%
  subset(minExpr>0)%>%ungroup()%>%select(Symbol)%>%
  distinct()

##now let's get the LVs that are selectd by the RF model
top.lvs<-synTableQuery("SELECT * from syn21318452")$asDataFrame()%>%
  select(-c(ROW_ID,ROW_VERSION))

top.lvs<-as.data.frame(apply(top.lvs,2,function(x) gsub('^Y$','X',gsub('^N$','',x))))

mp_res<-synTableQuery("SELECT * FROM syn21046991")$asDataFrame()%>%
  filter(isCellLine != "TRUE")%>%
  subset(latent_var%in%top.lvs$LatentVar)%>%
  select(latent_var,id,value,specimenID,tumorType,modelOf,diagnosis)

expr.vars<-subset(all.vars,Hugo_Symbol%in%expr.genes$Symbol)

samps<-intersect(mp_res$specimenID,expr.vars$specimenID)

mp_res<-mp_res%>%
  subset(specimenID%in%samps)#

data.with.var<-mp_res%>%
  left_join(expr.vars,by='specimenID')

tab<-data.with.var

top.genes=tab%>%#group_by(tumorType)%>%
  mutate(numSamps=n_distinct(specimenID))%>%
      group_by(Hugo_Symbol)%>%
    mutate(numMutated=n_distinct(specimenID))%>%
    ungroup()%>%
  subset(numMutated>1)%>%
      subset(numMutated<(numSamps-1))%>%
  select(tumorType,Hugo_Symbol,numSamps,numMutated)%>%distinct()

##first spread the WT/Mutated values
vals<-tab%>%subset(Hugo_Symbol%in%top.genes$Hugo_Symbol)%>% #only look at those genes mutated
    mutate(mutated=ifelse(is.na(IMPACT),'WT','Mutated'))%>%
  select(latent_var,tumorType,value,Hugo_Symbol,specimenID,mutated)%>%
  distinct()%>%
  spread(key=Hugo_Symbol,value='mutated',fill='WT')

##double check to make sure there are both mutated and unmutated values
counts<-vals%>%
  gather(key=gene,value=status,-c(latent_var,tumorType,value,specimenID))%>%
    select(latent_var,tumorType,value,gene,specimenID,status)%>%
    group_by(latent_var,gene)%>%
    mutate(numVals=n_distinct(status))%>%
    mutate(numSamps=n_distinct(specimenID))%>%
    subset(numVals==2)%>%ungroup()

#so now we have only
with.sig<-counts%>%ungroup()%>%#subset(gene%in%top.genes$Hugo_Symbol)%>%
    group_by(latent_var,gene)%>%
  mutate(pval=wilcox.test(value~status)$p.value)%>%ungroup()%>%
  group_by(latent_var)%>%
  mutate(corP=p.adjust(pval))%>%ungroup()%>%
  select(latent_var,gene,pval,corP)%>%distinct()

sig.vals<-subset(with.sig,corP<0.01)


mut.counts<-sig.vals%>%group_by(latent_var)%>%
    summarize(`Number Genes Mutated`=n_distinct(gene))

DT::datatable(mut.counts)
```

## Get metaviper correlations

We already calculated the metaviper correlations here
```{r metaviper, warning=FALSE,echo=FALSE}

##get correlated metaviper scores
metaviper_scores='syn21259610'
mtab<-synapser::synTableQuery(paste('SELECT * FROM syn21322923 where corVal>0.65'))$asDataFrame()%>%
  group_by(latent_var)%>%summarize(numMetaviperCorrelated=n_distinct(gene))%>%
  rename(`Proteins with correlated activity`=numMetaviperCorrelated)

DT::datatable(mtab)
```

## Immune correlations
Here we are seeking correlations between Immune subtypes and latent variables.

```{r immune cor,warning=FALSE,echo=FALSE}

#lastly the immune deconv scores
#get immune predictions
deconv_scores='syn20710536'

dtab<-synapser::synTableQuery(paste('select * from',deconv_scores))$asDataFrame()%>%
  rename(immScore='score')%>%subset(method!='xcell')

#get RF-selected latent variables
lvs<-synTableQuery("select * from syn21318452")$asDataFrame()$LatentVar

combined=dtab%>%select(c(cell_type,method,specimenID,immScore))%>%
  left_join(subset(mp_res,latent_var%in%lvs),by='specimenID')

corVals=combined%>%group_by(cell_type,latent_var,method)%>%summarize(corVal=cor(immScore,value,method = "spearman"))


corthresh=0.65

cor_cell_types=subset(corVals,abs(corVal)>corthresh)%>%
  ungroup()%>%rowwise()%>%
  mutate(`Immune Cell Prediction`=paste(method,cell_type,sep=':'))%>%
  ungroup()%>%group_by(latent_var)%>%
  mutate(numTypes=n_distinct(`Immune Cell Prediction`))%>%
#  subset(numTypes>1)%>%
  mutate(cell_types=paste(unique(`Immune Cell Prediction`),collapse=','))%>%
  ungroup()%>%
  select(latent_var,cell_types,numTypes)%>%unique()

#print(paste('we found',nrow(cor_cell_types),'latent variables/methods with some cell types with correlation greater than',corthresh))

DT::datatable(cor_cell_types)
```

## Merge data together

Now that we have all the three correlates calculated we want to summarize the latent variables by whether or not that have associated pathway data affiliated with them. This would likely become table 2 or 3 in the manuscript.

```{r, warning=FALSE}

final.tab<-top.lvs%>%
  rename(latent_var='LatentVar')%>%
  left_join(select(cor_cell_types,c(latent_var,`Immune Signatures`='numTypes')),by='latent_var')%>%
  left_join(mtab,by='latent_var')%>%
  left_join(mut.counts,by='latent_var',fill=0)
 

#to browse in the markdown
DT::datatable(final.tab)

#to cut/paste into the manuscript
stab<-synapser::synBuildTable(final.tab,parent='syn21046734',name='RF-selected variables with correlates')
synapser::synStore(stab)
```
