---
title: "Tumor type association with latent variable correlated viper protein expression in NF tumors"
author: "Robert Allaway"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```


Here, we are evaluating the 'druggability' of viper proteins that are well correlated with variables. 
The plots at the end of this are labeled by LVs, but the enrichment is actually based on the list of viper proteins that have a correlation of 0.5 or greater with LV expression across all tumors.

# Import packages

First, import packages to process and plot the data. 

```{r message=FALSE, warning=FALSE}
library(dplyr)
library(tidyr)
library(purrr)
library(synapser)
synLogin()
library(clusterProfiler)

viper_res <- synTableQuery('SELECT * FROM syn21260871 where numSamps = 77')$filepath %>% readr::read_csv()

drug_targets <- feather::read_feather(synGet('syn20700199')$path)

drug_targets <- drug_targets %>% 
  filter(mean_pchembl > 7) %>% 
  mutate(gene= hugo_gene) %>% 
  select(gene, std_name) 
```

# Significance testing

First, create a list of all druggable genes. We're using the drug-target explorer data, for any gene for which there is drug-target relationship with a mean_pchembl value >7, which corresponds to 1 uM. This gives us a rough approximation of druggable targets in the human genome but could be a bit under conservative.

Then, we'll select the viper proteins that are 0.5 or greater correlation with latent variable expression, by latent variable (so each group of genes is based on a single latent variable expression, where each gene in the group is a highly-correlated viper protein)

Then, we'll perform a weighted Kolmogorov-Smirnov test using the `GSEA` function from the `clusterProfiler` package to assess whether any of the viper genes are enriched in the universe of druggable genes. Here, we're treating each LV as the ranked list of viper genes (for standard GSEA, this would be the differentially expressed genes, for example) and treating the list of all drug targets as the gene set we're looking for enrichment in. 


```{r message=FALSE, warning=FALSE}

term2gene <- drug_targets %>% 
  mutate(term = 'all_drugs') %>% 
  select(term, gene) %>% 
  distinct()

tidy <- viper_res %>% 
  group_by(latent_var) %>% 
  filter(corVal > 0.65) %>% #only positively correlated genes >0.65 or so...this is arbitrary
  top_n(15, corVal) %>% ##NB number 16 from Sara has a max of 15 - ie only top 15 if more than 15 greater than 0.65 cor
  filter(any(gene %in% term2gene$gene)) %>% ##remove LVs that have no overlap if any exist
  nest() %>% 
  mutate(data = lapply(data, function(x){
    geneList <- x$corVal 
    names(geneList) <- as.character(x$gene) 
    geneList <- sort(geneList, decreasing = TRUE)
    geneList
  })) ##nest gene lists, looking at top 5% of genes in each LV 

tidy_2 <- tidy %>% ###run GSEA on nested gene lists vs universe of "druggable genes" 
  mutate(data = sapply(data, function(x){
   tryCatch(GSEA(geneList = x, TERM2GENE = term2gene), error=function(err) NA)
 }))

tidy_3 <- tidy_2 %>%  ##store plots and data frames for LVs with significant results
  mutate(plot = lapply(data, function(x){
    if(nrow(x)>0){
      plot <- gseaplot(x, geneSetID = 1)
    }else{
        plot <- NA
    }
    return(plot)
    })) %>% 
    mutate(res = lapply(data, function(x){
    if(nrow(x)>0){
      res <- x@result
    }else{
        res <- NA
    }
    return(res)
    }))

tidy_filt <- tidy_3 %>% 
  filter(!is.na(plot)) %>% 
  select(-data)

```

Then we simply look for LVs where we got a significant (unadjusted) result, and then plot the results. There are `r nrow(tidy_filt)` LVs significantly enriched for drug targets (again, when considering the 95th percentile of genes in the LV). 

Here's a table of all results. p.adjust shows nothing as being significantly enriched, but I don't know enough about how the `GSEA` function does the p-value calculation to know if this makes much sense in this context. E.g. it's pretty clear that there are many druggable targets in each of the LVs in the table (see plots further down), but after adjustment nothing is signficant. One possibility is that the "gene set" of druggable targets is too large at about ~1500 genes? 

```{r}

apply(tidy_filt, 1 , function(x){

  res <- x[['res']] %>%  #we need to fix p.adj because of the way this was run
    mutate(p.adjust = p.adjust(pvalue, n = nrow(tidy))) %>% 
    mutate(lv = x[['latent_var']]) %>% 
    select(lv,setSize, enrichmentScore, NES,pvalue,p.adjust,
           leading_edge, core_enrichment)
  
  res
  
}) %>% bind_rows() %>% DT::datatable()


```

Let's plot these results. 

```{r}

apply(tidy_filt, 1, function(x){

  plt <- x[['plot']]
  
  res <- x[['res']] %>%  #we need to fix p.adj because of the way this was run
    mutate(p.adjust = p.adjust(pvalue, n = nrow(tidy))) %>% 
    mutate(lv = x[['latent_var']])
  
  tab_1 <- select(res,lv,setSize, enrichmentScore, NES,pvalue,p.adjust) %>% 
    ggpubr::ggtexttable(theme = ggpubr::ttheme(base_size = 10))
  
  tab_2 <- select(res,leading_edge, core_enrichment) %>% 
    mutate(core_enrichment = stringr::str_trunc(core_enrichment, width = 30)) %>% 
    ggpubr::ggtexttable(theme = ggpubr::ttheme(base_size = 10))
  
  tab_3 <- viper_res %>% 
    filter(latent_var %in% res$lv) %>% 
    filter(corVal > 0.5) %>% #only positively correlated genes >0.5 or so...this is arbitrary
    filter(any(gene %in% term2gene$gene)) %>% 
    ggpubr::ggtexttable(theme = ggpubr::ttheme(base_size = 10))

 
   cowplot::plot_grid(plt, tab_1, tab_2,tab_3, nrow=4,ncol=1, rel_heights = c(5,1,1))
})



```
Here are the highly correlated viper proteins for each of the above LVs. 

```{r}

apply(tidy_filt, 1, function(x){

  plt <- x[['plot']]
  
  res <- x[['res']] %>%  #we need to fix p.adj because of the way this was run
    mutate(p.adjust = p.adjust(pvalue, n = nrow(tidy))) %>% 
    mutate(lv = x[['latent_var']])
  
  print(viper_res %>% 
    filter(latent_var %in% res$lv) %>% 
    filter(corVal > 0.5))
  
  # cowplot::plot_grid(plt, tab_1, tab_2,tab_3, nrow=4,ncol=1, rel_heights = c(5,1,1))
})



```

```{r}

sessionInfo()

```
