---
title: "Tumor type association with latent variable correlated viper protein expression in NF tumors _ drugs only"
author: "Robert Allaway"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```


Here, we are evaluating the 'druggability' of viper proteins that are well correlated with variables. 
The plots at the end of this are labeled by LVs, but the enrichment is actually based on the list of viper proteins that have a are correlated with these LVs. 


For this analysis we only care about drugs that are _approved_ for use in humans. We use the DrugCentral list of drugs to filter this out. 


# Import packages

First, import packages to process and plot the data. 

```{r message=FALSE, warning=FALSE}
library(dplyr)
library(tidyr)
library(purrr)
library(synapser)
synLogin()
library(clusterProfiler)
library(enrichplot)
# library(ggupset)
library(ggplot2)
library(cowplot)

drugcentral_structures <- synTableQuery("SELECT InChIKey as inchikey, INN as name FROM syn21446643", 
                                        includeRowIdAndRowVersion = F)$filepath %>% 
  readr::read_csv()

viper_res <- synTableQuery('SELECT * FROM syn21260871 where numSamps = 77', includeRowIdAndRowVersion = F)$filepath %>% 
  readr::read_csv()

dtex_structures <- synTableQuery('SELECT inchikey, internal_id FROM syn17090819', includeRowIdAndRowVersion = F)$filepath %>% 
  readr::read_csv()

dtex_targets <- feather::read_feather(synGet('syn20700199')$path)

dtex_targets <- dtex_targets %>% 
  filter(mean_pchembl > 7) %>% 
  mutate(gene= hugo_gene) %>% 
  left_join(dtex_structures) %>% 
  inner_join(drugcentral_structures)

```


# Significance testing

First, create a list of all druggable genes. We're using the drug-target explorer data, for any gene for which there is drug-target relationship with a mean_pchembl value >7, which corresponds to 1 uM. This gives us a rough approximation of druggable targets in the human genome but could be a bit under conservative.

Then, we'll select the viper proteins that are 0.65 or greater correlation with latent variable expression, by latent variable (so each group of genes is based on a single latent variable expression, where each gene in the group is a highly-correlated viper protein)- this is basd on the threshold in notebook 16. 

Then, we'll perform a weighted Kolmogorov-Smirnov test using the `GSEA` function from the `clusterProfiler` package to assess whether any of the viper genes are enriched in the universe of druggable genes. Here, we're treating each LV as the ranked list of viper genes (for standard GSEA, this would be the differentially expressed genes, for example) and treating the list of all drug targets as the gene set we're looking for enrichment in. 


```{r message=FALSE, warning=FALSE}

term2gene <- dtex_targets %>% 
  mutate(term = name) %>% 
  select(term, gene) %>% 
  distinct()

tidy <- viper_res %>% 
  group_by(latent_var) %>% 
  nest() %>% 
  mutate(data = lapply(data, function(x){
    geneList <- x$corVal 
    names(geneList) <- as.character(x$gene) 
    geneList <- sort(geneList, decreasing = TRUE)
    geneList
  })) ##nest gene lists, looking at top 5% of genes in each LV 

res <- parallel::mclapply(tidy$data, function(x){
  GSEA(geneList = x, TERM2GENE = term2gene)
}, mc.cores = parallel::detectCores())

names(res) <- tidy$latent_var

##determine if no results for a given LV
empty_idx <- lapply(res, function(x){
  if(nrow(x@result)==0){
    TRUE
  }else{
    FALSE
  }
})

##remove if no results
res_complt <- res[empty_idx==F]


```

time to do ALL OF THE PLOTS!

First, dotplot:

```{r}

targ_counts <- dtex_targets %>% 
  group_by(name) %>% 
  summarize(count = n())

plots <- lapply(names(res_complt), function(x){ 
  foo <- res_complt[[x]]
  
  show <- c(foo$Description[order(foo$enrichmentScore, decreasing=TRUE)][1:10],
            foo$Description[order(foo$enrichmentScore, decreasing=FALSE)][1:10])

  plt1 <- dotplot(foo, showCategory = show, x = "count", orderBy = "Count") + 
    theme_bw() + theme(legend.position="left") +
    scale_y_discrete(label = function(y) stringr::str_trunc(y, 30))

 
  targets <- targ_counts %>% filter(name %in% show)
             
  bar <- fortify(foo, showCategory = show, split = NULL) %>% 
    full_join(targets, by = c("ID"="name")) %>% 
    mutate(x=eval(parse(text="Count")))

  idx <- order(bar[['Count']], decreasing = TRUE)
  bar$Description <- factor(bar$Description, rev(unique(bar$Description[idx])))

  plt2 <- ggplot(data = bar) +
     geom_bar(aes(x=Description, y = count), stat = "identity") + 
     coord_flip() + 
    theme_bw() +
    theme(axis.title.y=element_blank(),
        axis.text.y =element_blank(),
        axis.ticks.y=element_blank()) +
    labs(y = "Total number of targets")
    
  
  legend <- cowplot::get_legend(
  # create some space to the left of the legend
  plt1 + theme(legend.box.margin = margin(0, 0, 0, 12))
)
  
  plots <- cowplot::plot_grid(plt1 + theme(legend.position = 'none'),
                     plt2,
                     legend,
                     ncol = 3,
                     rel_widths = c(3,1,1), align='h', axis = 'left')
  
  title <- ggdraw() + 
  draw_label(
    x,
    fontface = 'bold',
    x = 0,
    hjust = 0
  ) +
  theme(
    plot.margin = margin(0, 0, 0, 7)
  )

  plot_grid(
  title, plots,
  ncol = 1,
  # rel_heights values control vertical title margins
  rel_heights = c(0.1, 1)

  )
})

plots

```

Then, network plots:

```{r}
plots <- lapply(names(res_complt), function(x){ 
  foo <-res_complt[[x]]
  show <- c(foo$Description[order(foo$NES, decreasing=TRUE)][1:10],
            foo$Description[order(foo$NES, decreasing=FALSE)][1:10])
  plt2 <- cnetplot(res_complt[[x]], categorySize="pvalue", showCategory = show,
                    foldChange=tidy$data[tidy$latent_var==x] %>% unlist) +
     ggplot2::ggtitle(x)
})

plots

```


Then a heatmap of the top 20 by p value: 

```{r}
plots <- lapply(names(res_complt), function(x){ 
    foo <-res_complt[[x]]
  show <- c(foo$Description[order(foo$NES, decreasing=TRUE)][1:10],
            foo$Description[order(foo$NES, decreasing=FALSE)][1:10])
  plt3 <- heatplot(res_complt[[x]], showCategory = show) + ggplot2::ggtitle(x)
})

plots

```

Then a ridgeplot of the top 30 - this function doesn't let you filter by name, so it's just a selection of 30 most significant in both directions - this is good for demonstrating directionality. Some of these are VIPER proteins that are negatively, not positively, correlated with the LVs:

```{r}

plots <- lapply(names(res_complt), function(x){ 
    foo <-res_complt[[x]]
  show <- c(foo$Description[order(foo$NES, decreasing=TRUE)][1:10],
            foo$Description[order(foo$NES, decreasing=FALSE)][1:10])
  plt4 <- ridgeplot(res_complt[[x]], showCategory = 30) + ggplot2::ggtitle(x)
})

plots

```

<!-- Then an upsetplot becaue of course we have to try that out: (NOTE, got rid of these due to knitting issue)--> 


<!-- ```{r} -->
<!-- plots <- lapply(names(res_complt), function(x){  -->
<!--     foo <-res_complt[[x]] -->
<!--   show <- c(foo$Description[order(foo$NES, decreasing=TRUE)][1:10], -->
<!--             foo$Description[order(foo$NES, decreasing=FALSE)][1:10]) -->
<!--   plt5 <- enrichplot::upsetplot(res_complt[[x]], showCategory = show) -->

<!-- }) -->

<!-- plots -->

<!-- ``` -->

<!-- Then a combined GSEA plot of the top ten by enrichment score: -->

<!-- ```{r} -->
<!-- plots <- lapply(names(res_complt), function(x){  -->
<!--     foo <-res_complt[[x]] -->
<!--   show <- c(foo$Description[order(foo$NES, decreasing=TRUE)][1:10], -->
<!--             foo$Description[order(foo$NES, decreasing=FALSE)][1:10]) -->
<!--   plt6 <- enrichplot::gseaplot2(res_complt[[x]], geneSetID = show) -->
<!-- }) -->

<!-- plots -->

<!-- ``` -->

table of all significant results

```{r}

tab <- lapply(names(res_complt), function(x){ 
  foo <- res_complt[[x]]@result %>% 
    mutate(latent_var = x) %>% 
    select(latent_var, everything())
}) %>% bind_rows()

# lv_drug_tab <- synBuildTable("Latent Variable Drug Set Enrichment Analysis", "syn21046734", tab)
# synStore(lv_drug_tab)

DT::datatable(tab)
```


```{r}
sessionInfo()
```